### PR TITLE
MF-M07: Unblock receiver states after finalized escrow operations

### PR DESCRIPTION
**Pull request by philanton**
## Summary
- Replace the `lastSignedState.EscrowChannelID != nil` guard in `issueTransferReceiverState` (`channel_v1`) and `issueReleaseReceiverState` (`app_session_v1`) with a new `EnsureNoOngoingEscrowOperation(wallet, asset)` store method.
- Predicate keys on the latest signed transition type instead of the raw `EscrowChannelID` field: `EscrowLock` / `MutualLock` always block; `EscrowDeposit` / `EscrowWithdraw` block only when the on-chain escrow channel `state_version` hasn't caught up; everything else allows.
- Adds DB-store tests covering all branches (no prior state, non-escrow, lock-types, finalize synced/unsynced, unsigned).

## Background (audit MF-M07)
The previous guard rejected any receiver-side state whenever the receiver's latest signed state had a non-nil `EscrowChannelID`. The field is cleared by `State.NextState()` only when the previous transition was `EscrowDeposit` or `EscrowWithdraw`, so the signed state itself retains it after a finalized escrow flow. As a result, a user who completed a deposit/withdrawal could be unable to receive transfers or app-session releases until an unrelated off-chain action advanced past the finalized state.

The replacement matches the existing `EnsureNoOngoingStateTransitions` pattern (transition type + state.version vs channel.state_version) but is scoped to escrow concerns only, so home-deposit / home-withdrawal logic is untouched.

## Test plan
- [x] `go test ./nitronode/store/database/... -run EnsureNoOngoingEscrowOperation -v`
- [x] `go test ./nitronode/...`
- [x] `go test ./pkg/...`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  - Enhanced escrow operation validation to detect and block state transitions when escrow activities are in-flight, including locks, deposits, and withdrawals for wallet-asset pairs.
  - Strengthened operational safeguards by preventing state changes during pending escrow operations not yet finalized on-chain.
  - Improved consistency of escrow-related state management across payment channel and application session handlers.

### CODE REVIEW

**Review by ihsraham — APPROVED** [review]
approved for m-07 closure.

the fix removes the broad `EscrowChannelID != nil` block and uses the escrow-state check on both receiver paths. finalized deposit and withdrawal states can receive again after the escrow channel version has caught up, while real in-flight locks still block. that matches the finding and keeps the safer on-chain sync guard.

File: nitronode/store/database/db_store.go [hunk 1]
```diff
@@ -237,3 +237,65 @@ func (s *DBStore) EnsureNoOngoingStateTransitions(wallet, asset string) error {
...
+
+	case core.TransitionTypeEscrowDeposit:
+		if result.EscrowChannelVersion != nil && result.StateVersion != *result.EscrowChannelVersion {
+			return fmt.Errorf("escrow deposit finalization is still ongoing")
+		}
+
+	case core.TransitionTypeEscrowWithdraw:
+		if result.EscrowChannelVersion != nil && result.StateVersion != *result.EscrowChannelVersion {
+			return fmt.Errorf("escrow withdrawal finalization is still ongoing")
+		}
```
Comment (nksazonov) [inline nitronode/store/database/db_store.go:297]: **Behavioral gap: `EscrowChannelVersion == nil` silently allows instead of blocking.**

The condition `result.EscrowChannelVersion != nil && result.StateVersion != *result.EscrowChannelVersion` treats a nil version (LEFT JOIN found no row in `channels`) as finalized. If `s.escrow_channel_id` is set on the signed state but the corresponding row is absent from `channels` (e.g. node restart before the channel record was synced, or DB inconsistency), the on-chain status is unknown yet the function allows issuance of receiver states.

Suggested fix: invert the nil guard so a missing channel record is treated as pending: change to `if result.EscrowChannelVersion == nil || result.StateVersion != *result.EscrowChannelVersion` and return the finalization-in-progress error.

File: nitronode/store/database/db_store_test.go [hunk 2]
```diff
@@ -646,3 +646,209 @@ func TestDBStore_EnsureNoOngoingStateTransitions(t *testing.T) {
...
+		store := NewDBStore(db)
+		require.NoError(t, store.CreateChannel(homeChannel))
+		require.NoError(t, store.CreateChannel(newEscrowChannel(1)))
+
+		state := newSignedState(2, core.TransitionTypeEscrowLock, true)
+		state.UserSig = nil
+		state.NodeSig = nil
+		storeState(t, store, state)
+
+		err := store.EnsureNoOngoingEscrowOperation(wallet, asset)
```
Comment (nksazonov) [inline nitronode/store/database/db_store_test.go:1214]: **Missing test: `EscrowDeposit`/`EscrowWithdraw` when the escrow channel row is absent from `channels`.**

All existing sub-tests for deposit and withdrawal rely on a `channels` row being present. There is no coverage for the case where `escrow_channel_id` is set on the signed state but `CreateChannel` was never called for the escrow channel, which is precisely the path affected by the nil-version gap above.

Suggested fix: add sub-tests `"EscrowDeposit - escrow channel missing from DB - block"` and `"EscrowWithdraw - escrow channel missing from DB - block"` that store an escrow-type signed state without creating the escrow channel, then assert that `EnsureNoOngoingEscrowOperation` returns an error.

**Review by nksazonov — CHANGES_REQUESTED** [review]
The PR correctly addresses the MF-M07 audit finding. The `EscrowChannelID != nil` guard was replaced with a purpose-built `EnsureNoOngoingEscrowOperation()` method that correctly distinguishes unresolved escrow locks from finalized deposit/withdrawal states. The approach is more conservative than the audit recommended (it additionally requires on-chain version confirmation for deposit/withdraw states, rather than relying solely on `NextState()` semantics), which is a sound design choice. The test coverage is thorough for the sunny-path variants.

There is one behavioral gap in the new predicate and a corresponding missing test case that should be addressed before merging.

File: nitronode/store/database/db_store.go (hunk 1 above)
Comment (philanton) [reply nitronode/store/database/db_store.go:297]: Yeah, good catch. Fixed in [b91766d](https://github.com/layer-3/nitrolite/pull/735/changes/b91766d5e312c77acc35818a549481cffbd2e5ea) — flipped the nil guard to `== nil ||` so a missing `channels` row is treated as pending. Applied the same fix to `EnsureNoOngoingStateTransitions` since all 6 branches had the same gap.

File: nitronode/store/database/db_store_test.go (hunk 2 above)
Comment (philanton) [reply nitronode/store/database/db_store_test.go:1214]: Added in [b91766d](https://github.com/layer-3/nitrolite/pull/735/changes/b91766d5e312c77acc35818a549481cffbd2e5ea) — sub-tests `EscrowDeposit/EscrowWithdraw - escrow channel missing from DB - block` cover the gap. Also added the equivalent missing-channel-row tests for each branch in `EnsureNoOngoingStateTransitions` since the same nil-guard fix landed there.

**Review by nksazonov — APPROVED** [review]
Good work!
